core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/controller/promisehandlers/SkewDetectionHandler.scala

```diff
@@ -0,0 +1,150 @@
+    *
+    * @param worker
+    * @return
+    */
+  def isEligibleForSkewed(worker: ActorVirtualIdentity): Boolean = {
+    !skewedToHelperWorkerHistory.values.toList.contains(
+      worker
+    )
+  }
+
+  /**
+    * worker is eligible for free if it is being used in neither of the phases.
+    *
+    * @param worker
+    * @return
+    */
+  def isEligibleForHelper(worker: ActorVirtualIdentity): Boolean = {
+    !skewedToHelperWorkerHistory.keySet.contains(
+      worker
+    )
+  }
+
+  def passSkewTest(
+      skewedWorkerCand: ActorVirtualIdentity,
+      helperWorkerCand: ActorVirtualIdentity,
+      loads: mutable.HashMap[ActorVirtualIdentity, WorkerWorkloadInfo],
+      etaThreshold: Int,
+      tauThreshold: Int
+  ): Boolean = {
+    if (
+      loads(
+        skewedWorkerCand
+      ).dataInputWorkload / Constants.defaultBatchSize > etaThreshold && (loads(
+        skewedWorkerCand
+      ).dataInputWorkload / Constants.defaultBatchSize > tauThreshold + loads(
+        helperWorkerCand
+      ).dataInputWorkload / Constants.defaultBatchSize)
+    ) {
+      return true
+    }
+    false
+  }
+
+  /** *
+    * returns an array of (skewedWorker, freeWorker, whether state replication has to be done)
+    */
+  def getSkewedAndHelperWorkersEligibleForFirstPhase(
+      loads: mutable.HashMap[ActorVirtualIdentity, WorkerWorkloadInfo]
+  ): ArrayBuffer[(ActorVirtualIdentity, ActorVirtualIdentity, Boolean)] = {
+    val retPairs = new ArrayBuffer[(ActorVirtualIdentity, ActorVirtualIdentity, Boolean)]()
+    // Get workers in increasing load
+    val sortedWorkers = loads.keys.toList.sortBy(loads(_).dataInputWorkload)
+
+    for (i <- sortedWorkers.size - 1 to 0 by -1) {
+      if (isEligibleForSkewed(sortedWorkers(i))) {
+        // worker has been previously paired with some worker and that worker will be used again.
+        if (skewedToHelperWorkerHistory.keySet.contains(sortedWorkers(i))) {
+          if (
+            passSkewTest(
+              sortedWorkers(i),
+              skewedToHelperWorkerHistory(sortedWorkers(i)),
+              loads,
+              Constants.reshapeEtaThreshold,
+              Constants.reshapeTauThreshold
+            )
+          ) {
+            // build table has already been replicated
+            retPairs.append(
+              (sortedWorkers(i), skewedToHelperWorkerHistory(sortedWorkers(i)), false)
+            )
+          }
+        } else if (i > 0) {
+          breakable {
+            for (j <- 0 to i - 1) {
+              if (
+                isEligibleForHelper(sortedWorkers(j)) && passSkewTest(
+                  sortedWorkers(i),
+                  sortedWorkers(j),
+                  loads,
+                  Constants.reshapeEtaThreshold,
+                  Constants.reshapeTauThreshold
+                )
+              ) {
+                retPairs.append((sortedWorkers(i), sortedWorkers(j), true))
```

Comment:
understood this 3-tuple by your explanation in zoom. we definitely need some documentation or naming to make this explicit, otherwise we easily forgot what are those three objects represent for in the future. Can we create a case class of those three fields instead of using an arbitrary 3-tuple?

Reply:
I don't want to create a case-class for something that will only be needed in this portion of the code. But I understood your point in the zoom call. I added many comments after that. Also, I modified the code to write it in a way that makes it very explicit. Essentially what I did is that I defined the following variables at the very beginning when using them:
```
val currSkewedWorker = skewedAndHelper._1
val currHelperWorker = skewedAndHelper._2
val stateTransferNeeded = skewedAndHelper._3
```

and then only use these variables.